compiler/rustc_mir_transform/src/scalar_deref_prop.rs

```diff
@@ -0,0 +1,505 @@
+        let Some((_target, rvalue)) = stmt.kind.as_assign() else { bug!() };
+        assert_eq!(_target.as_local(), Some(local));
+
+        match rvalue {
+            // This is a copy, just use the value we have in store for the previous one.
+            // As we are visiting in `assignment_order`, ie. reverse postorder, `rhs` should
+            // have been visited before.
+            Rvalue::Use(Operand::Copy(place) | Operand::Move(place))
+            | Rvalue::CopyForDeref(place) => {
+                if let Some(rhs) = place.as_local() {
+                    let target = targets[rhs];
+                    if matches!(target, Value::Pointer(_)) {
+                        targets[local] = target;
+                    } else if matches!(ssa.assignments[rhs], Set1::One(_)) {
+                        targets[local] = Value::Pointer(tcx.mk_place_deref(rhs.into()));
+                    }
+                }
+            }
+            Rvalue::Ref(_, _, place) | Rvalue::AddressOf(_, place) => {
+                let mut place = *place;
+                if place.projection.first() == Some(&PlaceElem::Deref)
+                    && let Value::Pointer(target) = targets[place.local]
+                {
+                    place = target.project_deeper(&place.projection[1..], tcx);
+                }
+                assert_ne!(place.local, local);
+                if ssa.is_constant(place) {
+                    targets[local] = Value::Pointer(place);
+                }
+            }
+            // We do not know what to do, so keep as not-a-pointer.
+            _ => {}
+        }
+    }
+
+    debug!(?targets);
+
+    targets
+}
+
+/// Compute the set of locals that can be fully replaced.
+///
+/// We consider a local to be replacable iff it's only used in a `Deref` projection `*_local` or
+/// non-use position (like storage statements and debuginfo).
+fn fully_replacable_locals(ssa: &SsaLocals, copies: &IndexVec<Local, Local>) -> BitSet<Local> {
+    let mut replacable = BitSet::new_empty(ssa.direct_uses.len());
+
+    // First pass: for each local, whether its uses can be fully replaced.
+    for (local, &num_direct_uses) in ssa.direct_uses.iter_enumerated() {
+        if num_direct_uses == 0 {
+            replacable.insert(local);
+        }
+    }
+
+    // Consolidate to have a local iff all its copies are.
+    //
+    // `copies` defines equivalence classes between locals.  The `local`s that recursively
+    // move/copy the same local all have the same `head`.
+    for (local, &head) in copies.iter_enumerated() {
+        // If any copy is not replacable, then the head is not.
+        if !replacable.contains(local) {
+            replacable.remove(head);
+        }
+    }
+    for (local, &head) in copies.iter_enumerated() {
+        // If any copy is not replacable, then the head is not, then no copy is replacable.
+        if !replacable.contains(head) {
+            replacable.remove(local);
+        }
+    }
```

Comment:
Accidentally duplicated?

Reply:
No. The first loop removes the head if any duplicates is not in the bitset, and the second loop removes all duplicates if the head is not in the bitset.